src/actions/actions.ts

```diff
@@ -1812,53 +1812,108 @@ class ActionChangeChar extends BaseCommand {
 class MovementAWordTextObject extends BaseMovement {
   modes = [ModeName.Normal, ModeName.Visual];
   keys = ["a", "w"];
+  canBePrefixedWithCount = true;
 
   public async execAction(position: Position, vimState: VimState): Promise<IMovement> {
-    if (vimState.currentMode === ModeName.Visual && !vimState.cursorPosition.isEqual(vimState.cursorStartPosition)) {
-      // TODO: This is kind of a bad way to do this, but text objects only work in
-      // visual mode if you JUST entered visual mode
-      // TODO TODO: I just looked at this again and omg this is awful what was I smoking plz get rid of this asap
-      return {
-        start: vimState.cursorStartPosition,
-        stop: await new MoveWordBegin().execAction(position, vimState),
-      };
-    }
+    let start: Position;
+    let stop: Position;
 
     const currentChar = TextEditor.getLineAt(position).text[position.character];
 
-    // TODO(whitespace) - this is a bad way to do this. we need some sort of global
-    // white space checking function.
-    if (currentChar === ' ' || currentChar === '\t') {
-      return {
-        start: position.getLastWordEnd().getRight(),
-        stop: position.getCurrentWordEnd()
-      };
+    if (/\s/.test(currentChar)) {
+        start = position.getLastWordEnd().getRight();
+        stop = position.getCurrentWordEnd();
     } else {
-      return {
-        start: position.getWordLeft(true),
-        stop: position.getWordRight().getLeft()
-      };
+        stop = position.getWordRight().getLeftThroughLineBreaks();
+
+        if (stop.isEqual(position.getCurrentWordEnd())) {
+          start = position.getLastWordEnd().getRight();
+        } else {
+          start = position.getWordLeft(true);
+        }
     }
+
+    if (vimState.currentMode === ModeName.Visual && !vimState.cursorPosition.isEqual(vimState.cursorStartPosition)) {
+        start = vimState.cursorStartPosition;
+    }
+
+    return {
+      start: start,
+      stop: stop
+    };
   }
 
+  public async execActionWithCount(position: Position, vimState: VimState, count: number): Promise<Position | IMovement> {
+      let recordedState = vimState.recordedState;
+      let result: IMovement = {
+        start: undefined,
+        stop: undefined
+      };
+
+      if (count < 1) {
```

Comment:
It's slightly different with `BaseMovement#execActionWithCount`, especially the way we calculate `start` and `stop`.

In `BaseMovement#execActionWithCount`, we will assign the return value of each time Movement to `result`, which means both `start` and `stop` of `result` will be updated. This works perfectly when you run `2w`, we just move the `start` property twice forward.

While here, we are not just moving cursor, instead we are expanding the selection. We need to cache the `start` of the first movement and cache the `stop` of the last movement.

The only common part of these two methods are the `count` tweaking. But you are right, we can have a better way of this method to avoid confusing other maintainers.